### ISSUE TITLE
Closing a socket will hang the asyncio poller

### ISSUE THREAD

**Issue by pvanderlinden** [zeromq/pyzmq#1114]
When closing a socket the asyncio poller will crash, even when unregistering first: The whole coroutine will be broken, and couldn't find a way to actually catch the exception.

Version info:
```
libzmq-4.1.6
pyzmq-16.0.2
Python-3.5.4 |Continuum Analytics, Inc.| (default, Aug 14 2017, 13:26:58) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]
```

Program to demonstrate the issue:
```
import asyncio
import zmq.asyncio as zmq_asyncio
import zmq
import logging

logging.basicConfig(level=logging.DEBUG)

zmq_asyncio.install()
ctx = zmq_asyncio.Context()
s1 = ctx.socket(zmq.PULL)
s1.bind('inproc://1')
s2 = ctx.socket(zmq.PULL)
s2.bind('inproc://2')
poller = zmq_asyncio.Poller()
poller.register(s1, zmq.POLLIN)
poller.register(s2, zmq.POLLIN)

async def ping():
    s = ctx.socket(zmq.PUSH)
    s.connect('inproc://1')
    s.connect('inproc://2')
    while True:
        s.send_unicode('ping')
        await asyncio.sleep(2)

async def unregister():
    await asyncio.sleep(1)
    poller.unregister(s1)
    await asyncio.sleep(0.5)  # Increase this to "solve" the bug
    s1.close()
    print('unregister')

async def listen():
    while True:
        try:
            poll_events = await poller.poll(2000)
        except (asyncio.CancelledError, KeyboardInterrupt):
            break
        except Exception:
            logging.exception('Error polling for events')
            continue
        for socket, poll_event in poll_events:
            msg = await socket.recv_unicode()
            print(socket, msg)
        if not poll_events:
            print('no msg')

loop = asyncio.get_event_loop()
asyncio.ensure_future(ping())
asyncio.ensure_future(unregister())
try:
    loop.run_until_complete(listen())
except Exception:
    logging.exception('error running loop')
finally:
    ctx.destroy()
```
It is expected to keep printing ping after removing and closing the socket. Instead the poller will get an exception and the statement `poller.poll` will never end, traceback:

```
ERROR:asyncio:Exception in callback _AsyncPoller.poll.<locals>.on_poll_ready(<Future cancelled>) at /opt/anaconda3/envs/venv/lib/python3.5/site-packages/zmq/eventloop/future.py:88
handle: <Handle _AsyncPoller.poll.<locals>.on_poll_ready(<Future cancelled>) at /opt/anaconda3/envs/venv/lib/python3.5/site-packages/zmq/eventloop/future.py:88>
Traceback (most recent call last):
  File "/opt/anaconda3/envs/venv/lib/python3.5/asyncio/events.py", line 127, in _run
    self._callback(*self._args)
  File "/opt/anaconda3/envs/venv/lib/python3.5/site-packages/zmq/eventloop/future.py", line 91, in on_poll_ready
    if watcher.exception():
  File "/opt/anaconda3/envs/venv/lib/python3.5/asyncio/futures.py", line 306, in exception
    raise CancelledError
concurrent.futures._base.CancelledError
```

**Comment by minrk** [comment]
Indeed, the cancel hookup is incorrect. The code assumes that a cancelled Future will have CancelledError as a regular exception, but unlike other failures, `.exception()` *itself* raises CancelledError, so cancel must be handled specially.

Fixed by #1115.

**Comment by pvanderlinden** [comment]
Thanks, that should fix it indeed.
